### DISCUSSION TITLE
New database integration -- OceanBase

### DISCUSSION THREAD

**Discussion by LuckyPickleZZ** [dbeaver/dbeaver#13175] [DBeaver Development]
Hello, everyone. I have a request to integrate our open source database OceanBase（ [https://open.oceanbase.com](url) ）into DBeaver, like Clickhouse and MYSQL, etc. However, OceanBase has independent plugins. We need to do some development before integration.

Why do we do it?
DBeaver is an open-source, widely used database management tool with high usability. By integrating OceanBase into DBeaver, we can reduce the use cost of OceanBase. We hope to promote OceanBase with the help of DBeaver.
OceanBase already has many enterprise users. So by integrating OB, we can expand the number of users of DBeaver. And there also are many active developers in OB. If OB can be integrated into the DBeaver community, I believe we will make some contributions to the open-source community.
At present, to use OCeanBase, you must deploy ODC(OceanBase Developer Center). This increases the cost of maintenance and tool management. However, if OCeanBase can be integrated into DBeaver, the cost will be much lower for small organizations or individuals.

How do we do it?
We want to develop a separate plugin, but we need to rely on MySQL and Oracle plugins.
Although OceanBase is an independently developed database, it is compatible with Oracle and MYSQL. Users do not need to spend much time on data migration. In other words, OceanBase provides two modes of connection.
That is, two plugins are required. Users can choose OceanBase modes by creating MySQL or Oracle tenants, and the interface should be different.
To avoid duplicate code, I wanna introduce org.jkiss.dbeaver.ext.mysql and org.jkiss.dbeaver.ext.oracle packages into MANIFEST.MF of OceanBase plugin. I don't know if this is feasible.

So, here are two questions:
Can we add two plugins for OceanBase like this?
<img width="317" alt="image" src="https://user-images.githubusercontent.com/36879124/125403559-3cfc0800-e3e8-11eb-89f0-6a5f4b296edf.png">

Can we introduce the two plugins mentioned above into our MANIFEST.MF ?
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/36879124/125403710-65840200-e3e8-11eb-8e30-78ea995b309c.png">

If someone could answer these two questions, you can reply in this discussion or send it to my email liuyizhuo.lyz@alipay.com ~ Any suggestions or comments will also be welcomed.

PS: What’s OceanBase?
OceanBase is a native distributed relational database software developed by Alibaba and AntGroup, which is not based on any open source products. It achieves financial high availability on ordinary hardware. It has excellent horizontal expansion ability and is the first distributed database in the world to pass the TPC-C standard test, The scale of a single cluster is more than 1500 nodes. Our product has the characteristics of the original distribution, strong consistency, and highly compatible Oracle/MySQL. It bears the core link of Alipay 100% and runs steadily in the core system of dozens of banks, insurance companies, and other financial customers.

**Comment by LuckyPickleZZ** [comment]
@serge-rider Hello, could you please view this ?
My questions are:
    Can I add two plugins for two modes of one database?
    Can I introduce org.jkiss.dbeaver.ext.mysql and org.jkiss.dbeaver.ext.oracle packages into OceanBase plugin?

**Reply by serge-rider** [reply]
Hi @LuckyPickleZZ 

Sure, you can con tribute new extension to DBeaver Community. We usually merge PRs.
I'd suggest to create just one plugin - it may provide two different drivers (any number of driver actually).
You can add dependencies to any plugins which are already part of DBeaver (like MySQL or Oracle).

One limitation: you shouldn't include any big 3rd party dependencies (like spring, jna, etc). We try to keep community package as compact as possible.
You can use existing database extension plugins as example (e.g. Firebird, Clickhouse, Snowflake).

**Reply by LuckyPickleZZ** [reply]
Thanks very much for your reply. We may apply for PR in a month.

**Comment by LuckyPickleZZ** [comment]
Product Positioning
This development is based on Dbeaver for secondary development. The main purpose is to add an independent plugin supporting OceanBase database connection, which can support MySQL mode and Oracle mode. Users select tenant mode on the connection page, and then use different interfaces and functions according to different tenant modes. The two modes need to coexist in the same plugin, that is, only one DB driver is provided .

Module design
According to the feedback from the community, as long as we do not reference the third-party package, we can arbitrarily reference the existing code in the tool.
Based on this, the core design idea is: Based on the existing modules of Oracle, MySQL and generic, try to realize the interface and function of OB plugin by inheriting and rewriting methods without affecting the parent class; If it cannot be implemented by overriding methods or the parent class code must be modified, the inheritance is not used and a new class is added to implement ob code.
At present, only MySQL adaptation and part of Oracle adaptation have been done. Since ob Oracle tenant mode requires a lot of adaptation work, modules will continue to be added in the future.

New module
ValueHandler
    

> OceanbaseValueHandlerProvider

Auth
    

> OceanbaseAuthModelDatabaseNative

Plan

> OceanbasePlanAnalyzer
> OceanbasePlanJSON
> OceanbasePlanNodeJSON

Session
    

> OceanbaseOracleServerSession
>     OceanbaseOracleServerSessionManager
>     OceanbaseSessionEditor

Ui
    

> Activator
>     OceanbaseMessages
>     OceanbaseConnectionPage

Modified module
    

> OceanbaseDataSourceProvider(extends MySQLDataSourceProvider)
>     OceanbaseMySQLDataSource(extends MySQLDataSource)
>     OceanbaseOracleDataSource(extends OracleDataSource)

Plugin modify reference description
The extension points implemented include
    

> org.jkiss.dbeaver.dataSourceProvider
>     org.jkiss.dbeaver.dataTypeProvider
>     org.eclipse.ui.editors
>     org.jkiss.dbeaver.dataSourceProvider
>     org.jkiss.dbeaver.dataSourceView
>     org.jkiss.dbeaver.dataSourceAuth
>     org.jkiss.dbeaver.ui.propertyConfigurator

Except datasourceprovider, they are all internal classes of oceanbase plugin

DataSourceProvider
For DataSourceProvider, the base class is OceanbaseDataSourceProvider. However, in order to include the tree structure of MySQL and Oracle, the < tree >... < / tree > structure in the two plugins is directly referenced. The isMysqlMode() method in datasource is used to control the display and hiding. Users can access different interface structures in different modes.
